---
title: "VCF metrics"
output: rmdformats::robobook
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = FALSE)

library(dplyr)
library(here)
library(vroom)
library(knitr)
library(ggplot2)

sample_info <- vroom(here("sample_info/internal_external_sample_info.tsv"))
```

```{r main_vcf_metrics}
final_vcf_metrics_detail <- here("10_FinalVCF","output","cohort_biallelicSNPs_HardFiltered_WithMissingness_filtered.variant_calling_detail_metrics") %>% 
  read.delim(stringsAsFactors = F, comment.char = "#", dec = ",") %>% 
  inner_join(sample_info, by = c("SAMPLE_ALIAS" = "sample_id"))

final_vcf_metrics_summary <- here("10_FinalVCF","output","cohort_biallelicSNPs_HardFiltered_WithMissingness_filtered.variant_calling_summary_metrics") %>% 
  read.delim(stringsAsFactors = F, comment.char = "#", dec = ",") 
```

```{r pop_vcf_metrics}

breed_vcf_metrics_summary <- 
  list(large_white = here("10_FinalVCF","output","cohort_biallelicSNPs_HardFiltered_WithMissingness_filtered_large_white..variant_calling_summary_metrics") %>%
       read.delim(stringsAsFactors = F, comment.char = "#", dec = ",") ,
     
     landrace = here("10_FinalVCF","output","cohort_biallelicSNPs_HardFiltered_WithMissingness_filtered_landrace..variant_calling_summary_metrics") %>%
       read.delim(stringsAsFactors = F, comment.char = "#", dec = ","),
     
     pietrian = here("10_FinalVCF","output","cohort_biallelicSNPs_HardFiltered_WithMissingness_filtered_pietrain..variant_calling_summary_metrics") %>%
       read.delim(stringsAsFactors = F, comment.char = "#", dec = ","),
     
     duroc = here("10_FinalVCF","output","cohort_biallelicSNPs_HardFiltered_WithMissingness_filtered_duroc..variant_calling_summary_metrics") %>%
       read.delim(stringsAsFactors = F, comment.char = "#", dec = ","),
     
     european_wild_boar = here("10_FinalVCF","output","cohort_biallelicSNPs_HardFiltered_WithMissingness_filtered_euro_wild_boar..variant_calling_summary_metrics") %>%
       read.delim(stringsAsFactors = F, comment.char = "#", dec = ",")) %>% 
  bind_rows(.id = "breed")

```


# SNPs in cohort
```{r include=TRUE}
final_vcf_metrics_summary %>%
  dplyr::select(-contains("INDEL"), -contains("INS_DEL"), -contains("MULTIALLELIC"), -FILTERED_SNPS) %>% 
  kable(digits = 3)
```

# SNPs per breed
```{r, include=TRUE}

breed_vcf_metrics_summary %>% 
  dplyr::select(-contains("INDEL"), -contains("INS_DEL"), -contains("MULTIALLELIC"), -FILTERED_SNPS, -SNP_REFERENCE_BIAS, NUM_SINGLETONS) %>% 
  DT::datatable(width = '100%', options = list(scrollX = TRUE, dom = 't'), rownames = FALSE) 

```

# SNPs per sample
```{r, include=TRUE}

final_vcf_metrics_detail %>% 
  dplyr::select(SAMPLE_ALIAS, breed, TOTAL_SNPS) %>% 
  ggplot(data = ., aes(x = SAMPLE_ALIAS, y = TOTAL_SNPS)) +
    geom_bar(stat = "identity") +
    facet_grid(~breed, scales = "free_x")+
    ylab("Number Of SNPs") +
    xlab(NULL) +
    theme_bw() +
    theme(axis.text.x = element_blank(),
          axis.text.y = element_text(size = 10)) +
    scale_y_continuous(breaks = seq(0,4e6,5e5), labels = scales::scientific, expand = c(0,0)) +
    ggtitle(label = "Number of biallelic SNPs per sample (GT filtered VCF)")

```

