---
title: "Fst analysis - SOSFERT samples - PI vs LW"
output: 
  html_document:
    toc: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = TRUE, message = FALSE)

library(here)
library(vroom)
library(dplyr)
library(knitr)
library(ggpubr)
library(GenomicRanges)
library(WebGestaltR)
```

```{r import_data, include=FALSE} 

fst_win_pi_lw <- vroom(here("13_Fst/output/fst_win_SOSFERT_PT_vs_LW.windowed.weir.fst"))

fst_win_pi_lw$MEAN_FST %>% range()

fst_win_pi_lw$MEAN_FST %>% summary()

fst_win_pi_lw %>% dim() #95338     6

sum(fst_win_pi_lw$MEAN_FST > 0.5) 

```

```{r prepare_data} 

fst_win_pi_lw_prep <- fst_win_pi_lw %>% 
  # keep windows with at least 10 SNPs
  filter(N_VARIANTS >= 10) %>% 
  # calculate z-scores
  filter(CHROM %in% c(1:18,"X")) %>% 
  # label records corresponding to X
  mutate(is_x = (CHROM == "X")) %>% 
  # group by autosomes and X
  group_by(is_x) %>% 
  # standarized autosomes and X separately
  mutate(zFst = scale(MEAN_FST)[,1]) %>% 
  ungroup() %>% 
  mutate(CHROM = factor(CHROM, levels = c(1:18,"X")))

```

```{r gene_set} 

ssc_gene_set <- readRDS(here("ref_ensemble95/gene_set/ssc_gene_set.rds"))

ssc_gene_set_with_descr <- readRDS(here("ref_ensemble95/gene_set/ssc_gene_set_with_description.rds"))

```

```{r biomart_Setup, eval=FALSE}
library(biomaRt)

ensembl <- useEnsembl(biomart = "genes", dataset = "sscrofa_gene_ensembl", version = 95)

```


# Summary

Genomic mean differentiation is 0.15. 99th quantile windows have an Fst >= 0.59, but still this is not very extreme (zFst = 3.26). None of the windows have an Fst = 1

Considering that the 99.9th percentile has an Fst of ~5. This can be considred as a good threshold to define outlier regions.

```{r} 

fst_win_pi_lw_prep %>% 
  dplyr::select(CHROM, BIN_START, BIN_END, MEAN_FST, zFst) %>% 
  reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END")) %>% 
  group_by(variable) %>% 
  summarise(min = min(value), 
            mean = mean(value), 
            median = median(value), 
            q99 = quantile(value, 0.99),
            q99.5 = quantile(value, 0.995),
            q99.9 = quantile(value, 0.999),
            max = max(value)) %>% 
  kable(digits = 2)

```

# Windowed Fst distribution
```{r}  

ggarrange(
  
  fst_win_pi_lw_prep %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, MEAN_FST, zFst) %>% 
    reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END")) %>%
    ggplot(aes(x = value)) +
      geom_histogram(bins = 50) +
      facet_wrap(~variable, nrow = 2, scales = "free") +
      theme_bw(base_size = 12) +
      scale_x_continuous(breaks = scales::pretty_breaks(10)) +
      ylab(NULL) +
      xlab(NULL),
  
  fst_win_pi_lw_prep %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, MEAN_FST, zFst) %>% 
    reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END")) %>% 
    ggplot(aes(y = variable, x = value)) +
      geom_violin() +
      facet_wrap(~variable, nrow = 2, scales = "free") +
      theme_bw(base_size = 12) +
      theme(axis.text.y = element_blank()) +
      scale_x_continuous(breaks = scales::pretty_breaks(10)) +
      ylab(NULL) +
      xlab(NULL),
  
  nrow = 1,
  ncol = 2
  
)

```

# Genomic distribution

Dashed lines indicate percentiles 99th, 99.5th and 99.9th.

```{r make_manhattan_plots, eval = FALSE}  

# https://danielroelfs.com/blog/how-i-create-manhattan-plots-using-ggplot/

data_cum <- fst_win_pi_lw_prep %>% 
  mutate(center = (BIN_START + (BIN_END-BIN_START)/2 )) %>% 
  group_by(CHROM) %>% 
  summarise(max_bp = max(center)) %>% 
  mutate(bp_add = lag(cumsum(max_bp), default = 0)) %>% 
  dplyr::select(CHROM, bp_add)

fst_win_pi_lw_prep_manhattan <- fst_win_pi_lw_prep %>% 
  mutate(center = (BIN_START + (BIN_END-BIN_START)/2 )) %>% 
  inner_join(data_cum, by = "CHROM") %>% 
  mutate(bp_cum = center + bp_add)


axis_set <- fst_win_pi_lw_prep_manhattan %>% 
  group_by(CHROM) %>% 
  summarise(center = mean(bp_cum))

p_man_fst_autosomes <- fst_win_pi_lw_prep_manhattan %>% 
  ggplot(aes(x = bp_cum, y = MEAN_FST, color = CHROM)) +
  geom_point(alpha = 0.75) +
  scale_x_continuous(label = axis_set$CHROM, breaks = axis_set$center, expand = c(1/1000,1/1000)) +
  scale_y_continuous(expand = c(1/100,1/100), breaks = scales::pretty_breaks(n = 10)) +
  scale_color_manual(values = rep(c("black","darkgrey"), nrow(axis_set)) ) +
  labs(x = NULL) +
  theme_minimal(base_size = 12) +
  geom_hline(yintercept = quantile(fst_win_pi_lw_prep_manhattan$MEAN_FST, c(.99, .995, .999)), linetype = "dashed", color = "red", size = 1, alpha = 0.7) +
  theme(
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(face = "bold", size = 12, vjust = 3)
  )

p_man_z_fst_autosomes <- fst_win_pi_lw_prep_manhattan %>% 
  ggplot(aes(x = bp_cum, y = zFst, color = CHROM)) +
  geom_point(alpha = 0.75) +
  scale_x_continuous(label = axis_set$CHROM, breaks = axis_set$center, expand = c(1/1000,1/1000)) +
  scale_y_continuous(expand = c(1/100,1/100), breaks = scales::pretty_breaks(n = 10)) +
  scale_color_manual(values = rep(c("black","darkgrey"), nrow(axis_set)) ) +
  labs(x = NULL) +
  theme_minimal(base_size = 12) +
  geom_hline(yintercept = quantile(fst_win_pi_lw_prep_manhattan$zFst,  c(.99, .995, .999)), linetype = "dashed", color = "red", size = 1, alpha = 0.7) +
  theme(
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(face = "bold", size = 12, vjust = 3)
  )


png(here("00_analysis/figures/manhattan_genomewide_fst_pi_lw_sosfert_outliers.png"), res = 300, width = 4000, height = 2000, units = "px"); print(p_man_fst_autosomes); dev.off()
png(here("00_analysis/figures/manhattan_genomewide_z_fst_pi_lw_sosfert_outliers.png"), res = 300, width = 4000, height = 2000, units = "px"); print(p_man_z_fst_autosomes); dev.off()
```

```{r display_manhattan_plots}

knitr::include_graphics(c(here("00_analysis/figures/manhattan_genomewide_fst_pi_lw_sosfert_outliers.png"),
                        here("00_analysis/figures/manhattan_genomewide_z_fst_pi_lw_sosfert_outliers.png")))

```

# Outlier definition

Genes in/near regions of extreme Fst (outliers) were identified.

Outliers are defined as regions in the top 0.1% of the genomic distribution (99.9th percentile), plus all adjacent regions in the top 1% (99th percentile). 

Regions in the top1% not nearby top0.1% regions are not considered outliers.

# Number of outliers

```{r define_outlier_regions} 

cutoffs <- fst_win_pi_lw_prep$zFst %>% quantile(c(0.99, 0.999))

outliers_1pct <- fst_win_pi_lw_prep %>% 
  filter(zFst > cutoffs[1]) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>% 
  # merge adjacent outliers
  reduce(min.gapwidth = 1)

outliers_0.1pct <- fst_win_pi_lw_prep %>% 
  filter(zFst > cutoffs[2]) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>% 
  # merge adjacent outliers
  reduce(min.gapwidth = 1)

# keep top 1% windows near top 0.1% windows
outliers_1pct_near_outliers_0.1pct <- subsetByOverlaps(outliers_1pct, outliers_0.1pct, maxgap = 0)

# combine windows
outliers <- c(outliers_0.1pct, outliers_1pct_near_outliers_0.1pct) %>% sort()

``` 

Number of outlier windows
```{r}
outliers %>% as_tibble() %>% group_by(seqnames) %>% summarise(n_windows = n())
```

Number of ouflier regions merged
```{r}
outliers %>% reduce(min.gapwidth = 1) %>% as_tibble() %>% group_by(seqnames) %>% summarise(n_intervals = n())
```

Although the number of of outlier windows corresponds to 55, once adjacent windows (gap no larger than 1) are merged, only 10 genomic regions are generated.


# Genes in outlier regions

```{r find_genes} 

genes_outliers <- subsetByOverlaps(ssc_gene_set, outliers, maxgap = 50000)

#genes_outliers %>% as_tibble() %>% write.csv(here("00_analysis/data/genes_outliers_fst_dam_ewb.csv"))


```

```{r save_list_for_comparison_with_SOSFERT}

genes_outliers %>% saveRDS(here("00_analysis/data/genes_outliers_Fst_PI_LW_SOSFERT_samples.rds"))
```


There are `r  genes_outliers %>% length()` genes in outlier regions.

### Genes in outlier regions 
```{r display_genes}

genes_outliers %>%
  as_tibble() %>% 
  arrange(seqnames, start, end) %>% 
  DT::datatable(rownames = FALSE, options = list(pageLength = 20))

```

# Enrichment Analysis

Enrichment analyis was done using all `r  genes_outliers %>% length()` genes.

Significant results were not found

```{r function_run_WebGestaltR, eval = FALSE} 
#---------------------------
# function_run_WebGestaltR
#---------------------------

run_WebGestaltR <- function(
  interestGene, # vector of genes as input
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "sscrofa",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = 0.1,
    topThr = 100,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}

```

```{r run_WebGestaltR, eval = FALSE} 
run_WebGestaltR(
  interestGene=unique(genes_outliers$gene_id),
  projectName="GOBP",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_analysis/data/WebGestaltR_fst_PI_LW_SOSFERT_samples")
)


run_WebGestaltR(
  interestGene=unique(genes_outliers$gene_id),
  projectName="KEGG",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_analysis/data/WebGestaltR_fst_PI_LW_SOSFERT_samples")
)

```

# Gene annotations (GOBP, KEGG)
```{r add_gobp_to_outlier_genes, eval = FALSE}

#listAttributes(ensembl) %>% View()

genes_outliers_gobp <- getBM(attributes = c("ensembl_gene_id", "go_id", "name_1006", "namespace_1003"), 
                             filters = "ensembl_gene_id", 
                             values = unique(genes_outliers$gene_id), 
                             mart = ensembl)

genes_outliers_gobp <- full_join(
  genes_outliers %>% as_tibble() %>% dplyr::select(gene_name, gene_id, gene_biotype),
  genes_outliers_gobp %>% as_tibble() %>% filter(namespace_1003 == "biological_process"),
  by = c("gene_id"="ensembl_gene_id")
  ) %>% 
  dplyr::rename(go_name = name_1006, go_type = namespace_1003)
  
```

```{r add_KEGG_to_outlier_genes, eval = FALSE}
library(KEGGREST) #BiocManager::install("KEGGREST")

# convert ensembl ids to entrez ids
ensembl_to_entrez <- getBM(attributes = c("ensembl_gene_id", "entrezgene"),#"kegg_enzyme"),
                           filters = "ensembl_gene_id", 
                           values = unique(genes_outliers$gene_id),
                           mart = ensembl) 

ensembl_to_entrez <- ensembl_to_entrez %>% 
  filter(!is.na(entrezgene)) %>% 
  mutate(ssc_entrezgene_id = paste0("ssc:",entrezgene))

# genes to query
tmp <- unique(ensembl_to_entrez$ssc_entrezgene_id) #values: gene ensembl ids

# indixes to split data
# only 10 genes at a time are allowed
length(tmp)
query_df <- data.frame(start = seq(1, length(tmp), 10)) %>% mutate(end = start + 9)
head(query_df)
tail(query_df)


# query in batches of max 10 genes
KEGGs <- lapply(1:nrow(query_df), function(i){
  
  print(i)
  query_genes <- tmp[query_df$start[i]:query_df$end[i]]
  kegg_res <- keggGet(query_genes) %>% 
    lapply(function(x){ 
      data.frame(pathway_id = names(x$PATHWAY) , pathway = x$PATHWAY ) %>% 
        mutate(entrez_gene = x$ENTRY %>% as.integer()) %>% 
        dplyr::select(entrez_gene, everything())
      }) %>% 
    bind_rows()
  
  if(nrow(kegg_res) == 0){
    print("Neither gene matched KEGG terms")
  }else{
    kegg_res %>% 
      inner_join(ensembl_to_entrez, by = c("entrez_gene"="entrezgene")) 
  }
  }) 


genes_outliers_kegg <- KEGGs %>% 
  bind_rows() %>% 
  unique() %>% 
  as_tibble() %>% 
  dplyr::select(-ssc_entrezgene_id)
```

```{r combine_annotations_and_export, eval = FALSE}
genes_outliers_gobp_kegg <- full_join(genes_outliers_gobp, 
                                      genes_outliers_kegg, 
                                      by = c("gene_id"="ensembl_gene_id")) %>% 
  dplyr::select(gene_name, gene_id, entrez_gene, everything())

genes_outliers_gobp_kegg %>% 
  write.csv(here("00_analysis/data/SOSFERT_samples_FST_analysis_genes_outliers_gobp_kegg.csv"))
```

```{r import_gene_annotations, include=FALSE}

genes_outliers_gobp_kegg <- here("00_analysis/data/SOSFERT_samples_FST_analysis_genes_outliers_gobp_kegg.csv") %>% 
  read.csv() %>% 
  as_tibble() %>% 
  dplyr::select(-X)

genes_outliers_gobp_kegg
```

```{r combine_annotations_by_gene}
genes_outliers_gobp_kegg %>% 
  mutate(gene = ifelse(is.na(gene_name), gene_id, gene_name)) %>% 
  # add gene descriptions
  inner_join(
    ssc_gene_set_with_descr %>% 
      dplyr::select(gene_id, description), 
    by = "gene_id"
    ) %>% 
  # group by gene
  group_by(gene) %>% 
  summarise(gene_description = unique(description),
            GOBP = paste(unique(go_name), collapse = " / "),
            KEGG = paste(unique(pathway), collapse = " / ")) %>% 
  as_tibble() %>% 
  DT::datatable(rownames = FALSE)
```

# Genes of interest

After assessing gene ontology terms and KEGG pathways. The following set of genes is further characterized:

- ANKRD2: skeletal muscl tissue development (GOBP)

- CTSD: Estrogen signaling pathway (KEGG)

- IGF2: GOBPs related to muscle and reproduction.

- INS: KEGG includes "occyte meiosis", "ovarian steroidogenesis", "progesterone mediated oocyte maturation", "prolactin signalling pathway"; also KEGG for muscle "mTOR signaling pathway".

- PAPPA2: "Bone morphogenesis" (GOBP). It could be relevant for different is sizes/length between breeds (is there any??...need to check)

- PXT1: "peroxisomal testis enriched protein 1" (gene description)

- Genes involved in Wnt signaling, which a pathway involved in reproduction (i.e. folliculogenesis): FRAT1, FRAT2, SFRP5, TLE4

- TH: KEGG "prolactin signaling pathway"

- TNNI2 & TNN3: GOBP "skeletal muscle contraction"

More details are shown in the following table:

```{r}

genes_outliers_gobp_kegg_GOIs <- genes_outliers_gobp_kegg %>% 
  mutate(gene = ifelse(is.na(gene_name), gene_id, gene_name)) %>% 
  filter(
    gene %in% c("ANKRD2", "CTSD", "ENSSSCG00000035293", "INS", "PAPPA2", "PXT1", "FRAT1", "FRAT2", "SFRP5","TLE4", "TH", "TNNI2", "TNNT3")
    ) %>% 
  # add gene descriptions
  inner_join(
    ssc_gene_set_with_descr %>% 
      dplyr::select(seqnames, start, end, gene_id, description), 
    by = "gene_id"
    ) %>% 
  # group by gene
  group_by(seqnames, start, end, gene_id, gene) %>% 
  summarise(gene_description = unique(description),
            GOBP = paste(unique(go_name), collapse = " / "),
            KEGG = paste(unique(pathway), collapse = " / ")) %>% 
  as_tibble() %>% 
  arrange(seqnames, start, end) %>% 
  # locus col to make it nicer to display
  tidyr::unite("tmp", seqnames, start, sep = ":") %>% 
  tidyr::unite("gene_locus", tmp, end, sep = "-")   
```

```{r}
genes_outliers_gobp_kegg_GOIs %>% DT::datatable(rownames = FALSE)
```


### SNPeff predicted consequences

Checking for genes of interest carrying mutations that can affect the aminoacid chain sequence or gene splicing (HIGH/MODERATE SNPs). Also, the allele frequency of the mutations is shown in both breeds as well as the Fst.

The HIGH/MODERATE SNPs are mostly well differentiated between Pietrain and Large White. 

For the genes located in chr2, all the alternative alleles are fixed in Pietrain affecting the genes CTSD, TNNI2, TNNT3, which makes sense, as they are genes relevant for muscle function (except for CTSD). 

Other alleles well differentiated in favor to Pietrain affect ANKRD2, FRAT1, FRAT2. ANKRD2 is also relevant for muscle function, but since FRAT1 and FRAT2 are part of Wnt signaling pathway, I would have expected them to be affected by highly differentiated mutations occuring in Large White.

IGF1 (ENSSSCG00000035293) is affected by SNP that is almost fixed in LW and not found in PI, which is interesting as IGF1 causes muscle growth but it is also involved in fertility.

Finally, SFRP5 is affected by two missense mutation found only in LW (AF = 0.6). This is gene is part of wnt signaling.

```{r add_SNPeff_data}

genes_outliers_gobp_kegg_GOIs %>% 
  inner_join(
    here("16_SNPeff/output/cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered.ann.tab") %>% 
      vroom(),
    by=c("gene_id"="ANN[*].GENEID")
  ) %>% 
  as_tibble() %>% 
  dplyr::select(CHROM, POS, REF, ALT, `ANN[*].EFFECT`, `ANN[*].IMPACT`, gene_id, gene, gene_description) %>% 
  filter(!(`ANN[*].IMPACT` %in% c("MODIFIER", "LOW"))) %>%
  unique() %>% 
  # add alt allele frq PI
  inner_join(
    here("14_allele_frequencies/output/cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered_Pietrain_all_sites.frq2") %>% 
      vroom() %>% 
      dplyr::select(CHROM, POS, ALT_frq) %>% 
      dplyr::rename(ALT_frq_PI=ALT_frq),
    by = c("CHROM", "POS")
  ) %>% 
  # add alt allele frq LW
  inner_join(
    here("14_allele_frequencies/output/cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered_LargeWhite_all_sites.frq2") %>% 
      vroom() %>% 
      dplyr::select(CHROM, POS, ALT_frq) %>% 
      dplyr::rename(ALT_frq_LW=ALT_frq),
    by = c("CHROM", "POS")
  ) %>% 
  # add Fst information
  inner_join(
    here("13_Fst/output/fst_win_SOSFERT_PT_vs_LW.weir.fst") %>% 
      vroom(),
    by = c("CHROM", "POS")
  ) %>% 
  DT::datatable(rownames = FALSE)
```








# QTLs in outlier regions

Outliers were overlapped with QTLs.

QTLs were downloaded as a gff file from https://www.animalgenome.org/cgi-bin/QTLdb/SS/index ("All data by bp (on SS11.1 in gff format )") on April 1st.

```{r qtl_set, include=FALSE}

qtl_set <- readRDS(here("pigQTL/qtl_set.rds"))

qtl_set %>% 
  as.data.frame() %>% 
  group_by(qtl_class, seqnames) %>% 
  summarise(n=n()) %>% 
  ungroup() %>% 
  reshape2::dcast(qtl_class ~ seqnames, value.var = "n") %>% 
  knitr::kable()

```

Some QTLs correspond to very large intervals
```{r}

qtl_set %>% 
  width() %>% 
  as_tibble() %>% 
  ggplot(aes(x = value)) +
    geom_histogram() +
    theme_bw() +
    scale_x_continuous(breaks = scales::pretty_breaks(10)) 
```

But almost all intervals are short (Zooming in the most frequent interval width):
```{r}
qtl_set %>% 
  width() %>% 
  as_tibble() %>% 
  filter(value < 5e2) %>% 
  ggplot(aes(x = value)) +
    geom_histogram() +
    theme_bw() +
    scale_x_continuous(breaks = scales::pretty_breaks(10)) 

```

Check where those "short-most-frequent" QTLs occur
```{r}
qtl_set %>% 
  as_tibble() %>% 
  filter(width < 5e2) %>% 
  group_by(qtl_class, seqnames) %>% 
  summarise(n=n()) %>% 
  ungroup() %>% 
  reshape2::dcast(qtl_class ~ seqnames, value.var = "n") %>% 
  knitr::kable()
 
  
```



### Number of QTLs overlapping outlier regions 
```{r summary_qtl_class_dam} 

qtl_res <- subsetByOverlaps(qtl_set, outliers, maxgap = 0)

qtl_res %>% 
  as.data.frame() %>% 
  group_by(qtl_class, seqnames) %>% 
  mutate(seqnames = paste0("chr",seqnames),
         seqnames = factor(seqnames, levels = c(paste0("chr", 1:18)))) %>% 
  summarise(n_qtls = n()) %>% 
  reshape2::dcast(qtl_class ~ seqnames, value.var = "n_qtls") %>% 
  knitr::kable()
```

### Number of QTLs overlapping outliers and QTL names 
```{r summary_qtl_class_and_name_dam} 

qtl_res %>% 
  as.data.frame() %>% 
  dplyr::group_by(qtl_class, seqnames) %>% 
  mutate(seqnames = paste0("chr",seqnames)) %>% 
  summarise(n_qtls = n(), qtl_names = paste(unique(qtl_name), collapse = " / ")) %>% 
  DT::datatable(rownames = FALSE)

```

# Genes in QTLs

Find genes for those outliers mapped to QTL regions

```{r}
# Get outliers in QTLs
ov <- findOverlaps(query = outliers, subject = qtl_set, maxgap = 0)

outliers_in_qtl <- bind_cols(
  outliers[queryHits(ov)] %>% as_tibble() %>% dplyr::select(seqnames, start, end),
  qtl_set[subjectHits(ov)] %>% as_tibble() %>% dplyr::select(qtl_class, qtl_name)
  ) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE)

# Get genes for outliers in QTL
ovv <- findOverlaps(query = ssc_gene_set, subject = outliers_in_qtl, maxgap = 0)

genes_in_outliers_in_qtl <- bind_cols(
  ssc_gene_set[queryHits(ovv)] %>% as_tibble() ,
  outliers_in_qtl[subjectHits(ovv)] %>% as_tibble() %>% dplyr::select(qtl_class, qtl_name)
  ) %>% 
  mutate(gene = ifelse(is.na(gene_name), gene_id, gene_name)) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) 
```

### List of genes in outliers overlapping QTLs
```{r}
# tabulate genes by outlier type
genes_in_outliers_in_qtl %>% 
  as_tibble() %>% 
  dplyr::group_by(gene, qtl_class) %>% 
  summarise(n_qtls = n(), qtl_names = paste(unique(qtl_name), collapse = " / ")) %>% 
  DT::datatable(rownames = FALSE)
```

### Which genes are associated to reproduction?
```{r}
genes_in_outliers_in_qtl_repro <- genes_in_outliers_in_qtl %>% 
  as_tibble() %>% 
  filter(grepl("Repro",qtl_class))

```

```{r}
genes_in_outliers_in_qtl_repro %>% 
  dplyr::select(gene, gene_id) %>% 
  unique() %>% 
  inner_join(ssc_gene_set_with_descr, by = "gene_id") %>% 
  dplyr::select(gene, gene_id, description) %>% 
  DT::datatable(rownames = FALSE)
```


```{r}
genes_in_outliers_in_qtl_repro %>% 
  dplyr::group_by(gene, qtl_class) %>% 
  summarise(n_qtls = n(), qtl_names = paste(unique(qtl_name), collapse = " / ")) %>% 
  DT::datatable(rownames = FALSE)
```

### Which genes are associated to Meat and Carcass?
```{r}
genes_in_outliers_in_qtl_meat <- genes_in_outliers_in_qtl %>% 
  as_tibble() %>% 
  filter(grepl("Meat_and_Carcass",qtl_class))
```

```{r}
genes_in_outliers_in_qtl_meat %>% 
  dplyr::select(gene, gene_id) %>% 
  unique() %>% 
  inner_join(ssc_gene_set_with_descr, by = "gene_id") %>% 
  dplyr::select(gene, gene_id, description) %>% 
  DT::datatable(rownames = FALSE)
```

```{r}
genes_in_outliers_in_qtl_meat %>% 
  dplyr::group_by(gene, qtl_class) %>% 
  summarise(n_qtls = n(), qtl_names = paste(unique(qtl_name), collapse = " / ")) %>% 
  DT::datatable(rownames = FALSE)
```

### Some genes occur in both types of QTLs (repro and meat)
```{r}
inner_join(genes_in_outliers_in_qtl_repro %>% dplyr::select(gene_id) %>% unique(), 
           genes_in_outliers_in_qtl_meat %>% dplyr::select(gene_id) %>% unique(), 
           by = "gene_id") %>% 
  unique() %>% 
  inner_join(ssc_gene_set_with_descr, by = "gene_id") %>% 
  dplyr::select(gene_id, gene_name, description) %>% 
  DT::datatable()
```

### Genes in repro, not in meat QTLs ... all repro genes also overlap meat QTLs
```{r}
anti_join(genes_in_outliers_in_qtl_repro %>% dplyr::select(gene_id), 
           genes_in_outliers_in_qtl_meat %>% dplyr::select(gene_id), 
           by = "gene_id") %>% 
  unique() %>% 
  inner_join(ssc_gene_set_with_descr, by = "gene_id") %>% 
  dplyr::select(gene_id, gene_name, description) %>% 
  DT::datatable()
```

### Genes in meat, not in repro QTLs
```{r}
anti_join(genes_in_outliers_in_qtl_meat %>% dplyr::select(gene_id),
          genes_in_outliers_in_qtl_repro %>% dplyr::select(gene_id), 
          by = "gene_id") %>% 
  unique() %>% 
  inner_join(ssc_gene_set_with_descr, by = "gene_id") %>% 
  dplyr::select(gene_id, gene_name, description) %>% 
  DT::datatable()
```

