---
title: "Fst analysis - SOSFERT+External samples - PI vs LW"
output: 
  html_document:
    toc: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = TRUE, message = FALSE, warning = FALSE)

library(here)
library(vroom)
library(dplyr)
library(stringr)
library(knitr)
library(ggpubr)
library(GenomicRanges)
library(WebGestaltR)
```

```{r import_data, include=FALSE} 

fst_win_pi_lw <- vroom(here("13_Fst/output/fst_win_sosfert_external_PT_vs_LW.windowed.weir.fst"))

fst_win_pi_lw$MEAN_FST %>% range()

fst_win_pi_lw$MEAN_FST %>% summary()

fst_win_pi_lw %>% dim() #89647     6

sum(fst_win_pi_lw$MEAN_FST > 0.5) 

```

```{r prepare_data} 

fst_win_pi_lw_prep <- fst_win_pi_lw %>% 
  # keep windows with at least 10 SNPs
  filter(N_VARIANTS >= 10) %>% 
  # calculate z-scores
  filter(CHROM %in% c(1:18,"X")) %>% 
  # label records corresponding to X
  mutate(is_x = (CHROM == "X")) %>% 
  # group by autosomes and X
  group_by(is_x) %>% 
  # standarized autosomes and X separately
  mutate(zFst = scale(MEAN_FST)[,1]) %>% 
  ungroup() %>% 
  mutate(CHROM = factor(CHROM, levels = c(1:18,"X")))

```

```{r gene_set} 

ssc_gene_set <- readRDS(here("ref_ensemble95/gene_set/ssc_gene_set.rds"))

ssc_gene_set_with_descr <- readRDS(here("ref_ensemble95/gene_set/ssc_gene_set_with_description.rds"))

```

```{r biomart_Setup, eval=FALSE}
library(biomaRt)

ensembl <- useEnsembl(biomart = "genes", dataset = "sscrofa_gene_ensembl", version = 95)

```

# Summary

Genomic mean differentiation is 0.1. 99th quantile windows have an Fst >= 0.41, but still this is not very extreme (zFst = 3.44). None of the windows have an Fst = 1

Considering that the 99.9th percentile has an Fst of >5. This can be considered as a good threshold to define outlier regions.

```{r} 

fst_win_pi_lw_prep %>% 
  dplyr::select(CHROM, BIN_START, BIN_END, MEAN_FST, zFst) %>% 
  reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END")) %>% 
  group_by(variable) %>% 
  summarise(min = min(value), 
            mean = mean(value), 
            median = median(value), 
            q99 = quantile(value, 0.99),
            q99.5 = quantile(value, 0.995),
            q99.9 = quantile(value, 0.999),
            max = max(value)) %>% 
  kable(digits = 2)

```

# Windowed Fst distribution
```{r}  

ggarrange(
  
  fst_win_pi_lw_prep %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, MEAN_FST, zFst) %>% 
    reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END")) %>%
    ggplot(aes(x = value)) +
      geom_histogram(bins = 50) +
      facet_wrap(~variable, nrow = 2, scales = "free") +
      theme_bw(base_size = 12) +
      scale_x_continuous(breaks = scales::pretty_breaks(10)) +
      ylab(NULL) +
      xlab(NULL),
  
  fst_win_pi_lw_prep %>% 
    dplyr::select(CHROM, BIN_START, BIN_END, MEAN_FST, zFst) %>% 
    reshape2::melt(id.vars = c("CHROM","BIN_START","BIN_END")) %>% 
    ggplot(aes(y = variable, x = value)) +
      geom_violin() +
      facet_wrap(~variable, nrow = 2, scales = "free") +
      theme_bw(base_size = 12) +
      theme(axis.text.y = element_blank()) +
      scale_x_continuous(breaks = scales::pretty_breaks(10)) +
      ylab(NULL) +
      xlab(NULL),
  
  nrow = 1,
  ncol = 2
  
)

```


# Genomic distribution

Dashed lines indicate percentiles 99th, 99.5th and 99.9th.

```{r make_manhattan_plots, eval = FALSE}  

# https://danielroelfs.com/blog/how-i-create-manhattan-plots-using-ggplot/

data_cum <- fst_win_pi_lw_prep %>% 
  mutate(center = (BIN_START + (BIN_END-BIN_START)/2 )) %>% 
  group_by(CHROM) %>% 
  summarise(max_bp = max(center)) %>% 
  mutate(bp_add = lag(cumsum(max_bp), default = 0)) %>% 
  dplyr::select(CHROM, bp_add)

fst_win_pi_lw_prep_manhattan <- fst_win_pi_lw_prep %>% 
  mutate(center = (BIN_START + (BIN_END-BIN_START)/2 )) %>% 
  inner_join(data_cum, by = "CHROM") %>% 
  mutate(bp_cum = center + bp_add)


axis_set <- fst_win_pi_lw_prep_manhattan %>% 
  group_by(CHROM) %>% 
  summarise(center = mean(bp_cum))

p_man_fst_autosomes <- fst_win_pi_lw_prep_manhattan %>% 
  ggplot(aes(x = bp_cum, y = MEAN_FST, color = CHROM)) +
  geom_point(alpha = 0.75) +
  scale_x_continuous(label = axis_set$CHROM, breaks = axis_set$center, expand = c(1/1000,1/1000)) +
  scale_y_continuous(expand = c(1/100,1/100), breaks = scales::pretty_breaks(n = 10)) +
  scale_color_manual(values = rep(c("black","darkgrey"), nrow(axis_set)) ) +
  labs(x = NULL) +
  theme_minimal(base_size = 12) +
  geom_hline(yintercept = quantile(fst_win_pi_lw_prep_manhattan$MEAN_FST, c(.99, .995, .999)), linetype = "dashed", color = "red", size = 1, alpha = 0.7) +
  theme(
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(face = "bold", size = 12, vjust = 3)
  )

p_man_z_fst_autosomes <- fst_win_pi_lw_prep_manhattan %>% 
  ggplot(aes(x = bp_cum, y = zFst, color = CHROM)) +
  geom_point(alpha = 0.75) +
  scale_x_continuous(label = axis_set$CHROM, breaks = axis_set$center, expand = c(1/1000,1/1000)) +
  scale_y_continuous(expand = c(1/100,1/100), breaks = scales::pretty_breaks(n = 10)) +
  scale_color_manual(values = rep(c("black","darkgrey"), nrow(axis_set)) ) +
  labs(x = NULL) +
  theme_minimal(base_size = 12) +
  geom_hline(yintercept = quantile(fst_win_pi_lw_prep_manhattan$zFst,  c(.99, .995, .999)), linetype = "dashed", color = "red", size = 1, alpha = 0.7) +
  theme(
    legend.position = "none",
    panel.border = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(face = "bold", size = 12, vjust = 3)
  )


png(here("00_analysis/figures/manhattan_genomewide_fst_pi_lw_sosfert_external_outliers.png"), res = 300, width = 4000, height = 2000, units = "px"); print(p_man_fst_autosomes); dev.off()
png(here("00_analysis/figures/manhattan_genomewide_z_fst_pi_lw_sosfert_external_outliers.png"), res = 300, width = 4000, height = 2000, units = "px"); print(p_man_z_fst_autosomes); dev.off()
```

```{r display_manhattan_plots}

knitr::include_graphics(c(here("00_analysis/figures/manhattan_genomewide_fst_pi_lw_sosfert_external_outliers.png"),
                        here("00_analysis/figures/manhattan_genomewide_z_fst_pi_lw_sosfert_external_outliers.png")))

```

# Outlier definition

Genes in/near regions of extreme Fst (outliers) were identified.

Outliers are defined as regions in the top 0.1% of the genomic distribution (99.9th percentile), plus all adjacent regions in the top 1% (99th percentile). 

Regions in the top1% not nearby top0.1% regions are not considered outliers.

# Number of outliers

```{r define_outlier_regions} 

cutoffs <- fst_win_pi_lw_prep$zFst %>% quantile(c(0.99, 0.999))

outliers_1pct <- fst_win_pi_lw_prep %>% 
  filter(zFst > cutoffs[1]) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>% 
  # merge adjacent outliers
  reduce(min.gapwidth = 1)

outliers_0.1pct <- fst_win_pi_lw_prep %>% 
  filter(zFst > cutoffs[2]) %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>% 
  # merge adjacent outliers
  reduce(min.gapwidth = 1)

# keep top 1% windows near top 0.1% windows
outliers_1pct_near_outliers_0.1pct <- subsetByOverlaps(outliers_1pct, outliers_0.1pct, maxgap = 0)

# combine windows
outliers <- c(outliers_0.1pct, outliers_1pct_near_outliers_0.1pct) %>% sort()

``` 

Number of outlier windows
```{r}
outliers %>% as_tibble() %>% group_by(seqnames) %>% summarise(n_windows = n())
```

Number of ouflier regions merged
```{r}
outliers %>% reduce(min.gapwidth = 1) %>% as_tibble() %>% group_by(seqnames) %>% summarise(n_intervals = n())
```

# Genes in outlier regions

```{r find_genes} 

genes_outliers <- subsetByOverlaps(ssc_gene_set, outliers, maxgap = 50000)

```

```{r save_list_for_comparison_with_SOSFERT}

genes_outliers %>% saveRDS(here("00_analysis/data/genes_outliers_Fst_PI_LW_sosfert_external_samples.rds"))
```

There are `r  genes_outliers$gene_id %>% unique() %>% length()` genes in outlier regions.

### Genes in outlier regions 
```{r display_genes}

genes_outliers %>%
  as_tibble() %>% 
  arrange(seqnames, start, end) %>% 
  DT::datatable(rownames = FALSE, options = list(pageLength = 20))

```

# Enrichment Analysis

Enrichment analyis was done using all `r  genes_outliers$gene_id %>% unique() %>% length()` genes.

There were no significant results

```{r function_run_WebGestaltR, eval = FALSE} 
#---------------------------
# function_run_WebGestaltR
#---------------------------

run_WebGestaltR <- function(
  interestGene, # vector of genes as input
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "sscrofa",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = 0.1,
    topThr = 100,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}

run_WebGestaltR_hsa <- function(
  interestGene, # vector of genes as input
  projectName, # the suffix to which "Project_" is appended
  enrichDatabase, # the data base to query  
  outputDirectory # where to store the results
  ){
  res <- WebGestaltR(
    enrichDatabase = enrichDatabase,
    enrichMethod = "ORA",
    organism = "hsapiens",
    interestGene = interestGene,
    interestGeneType = "ensembl_gene_id",
    referenceSet = "genome",
    sigMethod = "fdr",
    fdrMethod = "BH",
    fdrThr = 0.1,
    topThr = 100,
    reportNum = 20,
    isOutput = TRUE,
    outputDirectory = outputDirectory,
    hostName = "http://www.webgestalt.org/",
    projectName = projectName
  )
  
  write.csv(
    res, 
    file.path(outputDirectory,paste0("Project_", projectName), paste0(enrichDatabase,"_sig_results.csv"))
  )
}

```

```{r run_WebGestaltR, eval = FALSE} 
run_WebGestaltR(
  interestGene=unique(genes_outliers$gene_id),
  projectName="GOBP",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_analysis/data/WebGestaltR_fst_PI_LW_SOSFERT_EXTERNAL_samples")
)


run_WebGestaltR(
  interestGene=unique(genes_outliers$gene_id),
  projectName="KEGG",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_analysis/data/WebGestaltR_fst_PI_LW_SOSFERT_EXTERNAL_samples")
)

```

Only significant results for KEGG, not GOBP.

```{r}
webgestalt_kegg_sig_res <- here("00_analysis/data/WebGestaltR_fst_PI_LW_SOSFERT_EXTERNAL_samples/Project_KEGG/pathway_KEGG_sig_results.csv") %>% 
  vroom()

webgestalt_kegg_sig_res %>% 
  DT::datatable(rownames = FALSE)
```

Genes enriching this pathway:
```{r}
ssc_gene_set_with_descr %>% 
  filter(
    gene_id %in% ( webgestalt_kegg_sig_res$userId %>% str_split(";") %>% unlist() )
      ) %>% 
  DT::datatable(rownames = FALSE)
```


# Enrichment Analysis with orthologous human genes

Outlier pig genes were mapped to their human orthologous and a new instance of WebGestalt was executed. 

```{r transform_ssc_to_hsa_ids, include=FALSE, eval = FALSE}

listAttributes(ensembl) %>% View()

vals <- genes_outliers$gene_id %>% unique()

vals %>% length() #86

ssc_outlier_genes_to_hsa <- getBM(attributes = c("ensembl_gene_id", "hsapiens_homolog_ensembl_gene"),  
                                  filters = "ensembl_gene_id", 
                                  values = vals, mart = ensembl)


ssc_outlier_genes_to_hsa$ensembl_gene_id %>% unique() %>% length() #86

ssc_outlier_genes_to_hsa %>% filter(hsapiens_homolog_ensembl_gene != "") %>% .$hsapiens_homolog_ensembl_gene %>% unique() %>% length() #79

# add gene infirmation and export

inner_join(ssc_outlier_genes_to_hsa, ssc_gene_set_with_descr, by = c("ensembl_gene_id"="gene_id")) %>% 
  as_tibble() %>% 
  unique() %>% 
  write.csv(here("00_analysis/data/WebGestaltR_fst_PI_LW_EXTERNAL_samples/Project_GOBP_hsa_ortho/ssc_to_hsa_orthologs_and_information_genes_fst_PI_LW_outliers_sosfert_external.csv"))

```

```{r run_WebGestaltR_hsa_ortho, eval = FALSE}

gns <- ssc_outlier_genes_to_hsa %>% filter(hsapiens_homolog_ensembl_gene != "") %>% .$hsapiens_homolog_ensembl_gene %>% unique()

run_WebGestaltR_hsa(
  interestGene=gns,
  projectName="GOBP_hsa_ortho",
  enrichDatabase="geneontology_Biological_Process",
  outputDirectory=here("00_analysis/data/WebGestaltR_fst_PI_LW_SOSFERT_EXTERNAL_samples")
)


run_WebGestaltR_hsa(
  interestGene=gns,
  projectName="KEGG_hsa_ortho",
  enrichDatabase="pathway_KEGG",
  outputDirectory=here("00_analysis/data/WebGestaltR_fst_PI_LW_SOSFERT_EXTERNAL_samples")
)

```

This time, significant results were found for GOBP and not for KEGG. Also immune related functions.

```{r import_significant_results}

webgestalt_ortho_gobp_sig_res <- here("00_analysis/data/WebGestaltR_fst_PI_LW_SOSFERT_EXTERNAL_samples/Project_GOBP_hsa_ortho/geneontology_Biological_Process_sig_results.csv") %>% 
  vroom()

webgestalt_ortho_gobp_sig_res %>% 
  DT::datatable(rownames = FALSE)

```


Genes enriching these pathway:
```{r}

lapply(1:nrow(webgestalt_ortho_gobp_sig_res), function(i){
  # arrange hsa-ensemble gene ids used for enrichment
  webgestalt_ortho_gobp_sig_res[,"userId"] %>% 
    unlist() %>% 
    str_split(";") %>% 
    unlist() %>% 
    as_tibble() %>% 
    magrittr::set_colnames("ensembl_gene_id") %>% 
    # add enriched term information
    mutate(
      term = webgestalt_ortho_gobp_sig_res[i,"description"] %>% unlist(),
      term_id = webgestalt_ortho_gobp_sig_res[i,"geneSet"] %>% unlist()
    ) %>%
    dplyr::select(term_id, term, ensembl_gene_id) %>% 
    # add ssc-gene ids
    inner_join(
      here("00_analysis/data/WebGestaltR_fst_PI_LW_EXTERNAL_samples/Project_GOBP_hsa_ortho/ssc_to_hsa_orthologs_and_information_genes_fst_PI_LW_outliers_sosfert_external.csv") %>% 
        vroom() %>% 
        dplyr::select(hsapiens_homolog_ensembl_gene, ensembl_gene_id) %>% 
        dplyr::rename(ssc_ensembl_gene_id=ensembl_gene_id),
      by = c("ensembl_gene_id"="hsapiens_homolog_ensembl_gene")
    ) %>% 
    # add gene description
    inner_join(
      ssc_gene_set_with_descr %>% 
        dplyr::select(gene_name, gene_id, description),
      by = c("ssc_ensembl_gene_id"="gene_id")
    ) %>% 
    unique() 
  }) %>% 
  bind_rows() %>% 
  group_by(term_id, term) %>% 
  summarise(genes = description %>% unique() %>% paste(collapse = " / ")) %>% 
  DT::datatable(rownames = FALSE)

```



