---
title: "ROHs analysis - SOSFERT samples"
output: 
  html_document:
    toc: 5
    code_folding: show
---

```{r libraries, include=FALSE}
#knitr::opts_chunk$set(echo = FALSE, include = TRUE, message = FALSE)
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)


library(vroom)
library(tidyverse)
library(GenomicRanges)
library(here)
library(knitr)
library(biomaRt)
library(rrvgo)

```

```{r sample_info}
sample_info <- vroom(here("sample_info/internal_external_sample_info.tsv"))
```


# Intro

Runs of homozygosity were computed for each sample of the SOSFERT sample set, using bcftools.

These samples were sequenced at high coverage of around 50x, so the VCF file had ~15M SNPs. These were used to compute ROH.

For more details check the corresponding script: in ``~/12_LD_ROH/scripts/12_bcftools_ROH_SOSFERT_vcf.sh``

```{r ROH_import_regions, results='hide'}

roh_dat <- vroom(here("12_LD_ROH/output/cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered.roh.table"), 
                 col_types = c(`[3]Chromosome`="c")) %>% 
  dplyr::rename(RG = `# RG`, sample = `[2]Sample`, chr = `[3]Chromosome`, start = `[4]Start`, end = `[5]End`, length_bp = `[6]Length (bp)`,
                n_markers = `[7]Number of markers`, qual_avg_fwd_bwd_phred = `[8]Quality (average fwd-bwd phred score)`) %>% 
  filter(RG != "# RG") %>% 
  # exclude x-chrom
  filter(chr != "X") %>% 
  inner_join(
    
    # include an ordered-naming for later when plotting all samples together (i.e. boxplot)
    sample_info %>% 
      filter(project == "SOSFERT") %>% 
      group_by(breed) %>% 
      mutate(breed = ifelse(breed == "Pietrain", "PI", breed),
             breed = ifelse(breed == "Large White", "LW", breed)) %>% 
      arrange(breed, sample_id), 

    by = c("sample" = "sample_id")
    )


dim(roh_dat) # 100140     10
```

```{r classify_rohs}
# Based on: "ROHs were separated into three size classes: 1) small (<100 Kbp), 2) medium (0.1 to 5 Mbp) and 3) large (>5 Mbp). "
# PLoS Genet. 2012;8(11):e1003100. doi: 10.1371/journal.pgen.1003100. Epub 2012 Nov 29. Regions of homozygosity in the porcine genome: consequence of demography and the recombination landscape
roh_dat_incl_class <- roh_dat %>% 
  mutate(roh_class = ifelse(length_bp < 1e5, "short", NA),
         roh_class = ifelse(length_bp >= 1e5 & length_bp <= 5e6, "medium", roh_class),
         roh_class = ifelse(length_bp > 5e6, "large", roh_class),
         roh_class = factor(roh_class, levels = c("short", "medium", "large")))

```

# ROH number and size summary
```{r roh_number_size_summary}

roh_dat_incl_class %>%
  group_by(breed, roh_class) %>% 
  summarise(n = n(), min_len = min(length_bp), max_len = max(length_bp)) %>% 
  arrange(roh_class, breed) %>% 
  kable(format.args = list(big.mark = ","))
  

```


# ROH length distribution by ROH-class
Based on: "ROHs were separated into three size classes: 1) small (<100 Kbp), 2) medium (0.1 to 5 Mbp) and 3) large (>5 Mbp). "
PLoS Genet. 2012;8(11):e1003100. doi: 10.1371/journal.pgen.1003100. Epub 2012 Nov 29. Regions of homozygosity in the porcine genome: consequence of demography and the recombination landscape

```{r roh_length_boxplots, fig.width=10}

roh_dat_incl_class %>% 
  ggplot(aes(x = length_bp, y = sample)) +
    geom_boxplot() +
    facet_grid(breed ~ roh_class, scales = "free") +
    theme_bw() +
    scale_x_continuous(breaks = scales::pretty_breaks(5)) +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.5, vjust = 0.5))

```

# Number of roh per class in each sample

```{r summarise_roh_counts}

roh_dat_incl_class %>% 
  group_by(sample,breed,roh_class) %>% 
  summarise(n=n()) %>% 
  reshape2::dcast(breed+sample ~ roh_class, value.var = "n") %>% 
  kable()
  
```

# Agreggate lengths (bp) of roh per class in each sample
```{r summarise_roh_aggreg_len}

roh_dat_incl_class %>% 
  group_by(sample,breed,roh_class) %>% 
  summarise(total_length=sum(length_bp)) %>% 
  reshape2::dcast(breed+sample ~ roh_class, value.var = "total_length") %>% 
  kable(format.args = list(big.mark = ","))

```

```{r vis_aggreagate_lengths}

roh_dat_incl_class %>% 
  group_by(breed, sample, roh_class) %>% 
  summarise(total_bp = sum(length_bp)) %>% 
  ungroup() %>% 
  group_by(breed, roh_class) %>% 
  summarise(mean = mean(total_bp/1e6), sd = sd(total_bp/1e6)) %>% 
  # add length of roh for reference
  mutate(
    roh_class = as.character(roh_class),
    roh_class = ifelse(roh_class == "large", "large (>5Mb)", roh_class),
    roh_class = ifelse(roh_class == "medium", "medium (100Kb-5Mb)", roh_class),
    roh_class = ifelse(roh_class == "short", "short (<100Kb)", roh_class),
    roh_class = factor(roh_class, levels = c("total","short (<100Kb)", "medium (100Kb-5Mb)", "large (>5Mb)"))
  ) %>% 
  ggplot(aes(x = breed, y = mean, fill = roh_class)) +
    geom_bar(stat = "identity", position = "dodge") +
    #geom_bar(stat = "identity", position = "stack") +
    geom_errorbar(aes(ymin = mean-sd, ymax = mean+sd), position =  position_dodge2(width = 0.5, padding = 0.5)) +
    theme_bw(base_size = 12) +
    xlab(NULL) +
    ylab("Mbp (mean +/- sd)") +
    ggtitle("Genome in ROH") +
    scale_fill_brewer(palette="Dark2") +
    scale_y_continuous(breaks = scales::pretty_breaks(10)) +
    theme(legend.title = element_blank(), plot.title = element_text(hjust = 0.5))


```


# Total fraction of genome in ROH (FROH)

The length of the genome covered by SNPs (denominator) is the result of:
"summing the distance between the first and the last available consecutive SNP on each chromosomal arm for each of the 22 autosomes" https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4573243/

Unfortunately, for pig centromeres are not mapped. The next best thing is to ignore per-arm territory and use the territory covered by all SNPs in the chromosome. 

```{r snp_territory}

genome_length <- here("10_FinalVCF/output/cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered.frq") %>% 
  vroom(col_select = c("CHROM","POS")) %>% 
  filter(CHROM %in% 1:18) %>% 
  group_by(CHROM) %>%
  summarise(territory_bp = (max(POS) - min(POS))+1 ) %>% 
  ungroup() %>% 
  summarise(territory_bp = sum(territory_bp)) %>% 
  as.numeric()
  
```

```{r chr_len}

len_assembly <- here("ref_ensemble95/Sus_scrofa.Sscrofa11.1.dna.toplevel.fa.fai") %>% vroom(col_names = FALSE) %>% summarise(sum(X2))

len_assembly_autosomes <- here("ref_ensemble95/Sus_scrofa.Sscrofa11.1.dna.toplevel.fa.fai") %>% 
  vroom(col_names = FALSE) %>% 
  filter(X1 %in% c(1:18)) %>% 
  summarise(sum(X2))

```

The genome length covered by SNPs (autosomes only) is `r genome_length`, whereas the length of the whole pig assembly is `r len_assembly`, only autosomes: `r len_assembly_autosomes`


```{r bp_in_roh_and_FROH_by_sample}

roh_dat_incl_class %>% 
  group_by(sample,breed) %>% 
  summarise(total_length_bp=sum(length_bp)) %>% 
  mutate(FROH = total_length_bp/genome_length) %>% 
  reshape2::melt(id.vars = c("sample","breed"), meassure.vars = c("total_length_bp", "FROH")) %>% 
  ggplot(aes(x = sample, y = value)) +
    geom_bar(stat = "identity") +
    facet_grid(variable ~ breed, scales = "free") +
    theme_bw(base_size = 12) +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.5, vjust = 0.5)) +
    scale_y_continuous(breaks = scales::pretty_breaks(10)) +
    ylab(NULL) + 
    xlab(NULL)

```


```{r per_sample_avg_genome_prop_as_roh}

# get genomic fraction
roh_dat_incl_class_genome_fraction <- roh_dat_incl_class %>% 
  mutate(genome_fraction = length_bp/genome_length)

# get total fraction per sample
total <- roh_dat_incl_class_genome_fraction %>% 
  group_by(breed, sample) %>% 
  summarise(total = sum(genome_fraction)) %>% 
  mutate(roh_class = "total") %>% 
  dplyr::select(breed, sample, roh_class, total)

# get total fraction by class per sample
total_by_class <- roh_dat_incl_class_genome_fraction %>% 
  group_by(breed, sample, roh_class) %>% 
  summarise(total = sum(genome_fraction))

# combine data
fractions <- bind_rows(total, total_by_class) %>% 
  #summarise by breed and class
  group_by(breed, roh_class) %>% 
  summarise(n(), mean = mean(total), sd = sd(total)) %>% 
  # add length of roh for reference
  mutate(
    roh_class = ifelse(roh_class == "large", "large (>5Mb)", roh_class),
    roh_class = ifelse(roh_class == "medium", "medium (100Kb-5Mb)", roh_class),
    roh_class = ifelse(roh_class == "short", "short (<100Kb)", roh_class)
  ) %>% 
  # arrange levels for plot
  mutate(roh_class = factor(roh_class, levels = c("total","short (<100Kb)", "medium (100Kb-5Mb)", "large (>5Mb)")))


errors_lower <- fractions$mean - fractions$sd
errors_upper <- fractions$mean + fractions$sd

ggplot(fractions, aes(x = breed, y = mean, fill = roh_class)) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_errorbar(aes(ymin = errors_lower, ymax = errors_upper), position =  position_dodge2(width = 0.5, padding = 0.5)) +
  theme_bw(base_size = 12) +
  xlab(NULL) +
  ylab("FROH (mean +/- sd)") +
  ggtitle("Mean FROH per breed per ROH class") +
  scale_fill_brewer(palette="Dark2") +
  scale_y_continuous(breaks = seq(0,1,0.05), expand = c(0.01,0.01)) +
  theme(legend.title = element_blank(), plot.title = element_text(hjust = 0.5))

```

# Private and shared ROHs

- Within breed common ROHs (breed-consensus) were detected using bedtools (see script: ``./12_LD_ROH/scripts/14_bedtools_multiIntersectBed.sh``).

- Based on biostars post: https://www.biostars.org/p/15303/

- ROHs are transformed to BED files for bedtools, afterwards intervals were reconverted into 1-start-fully-closed intervals.

- For more information see https://genome-blog.soe.ucsc.edu/blog/2016/12/12/the-ucsc-genome-browser-coordinate-counting-systems/

## Make a roh bed file for each sample
```{r transform_to_bed, eval=FALSE}

roh_dat_incl_class$sample %>% 
  unique() %>% 
  lapply(function(s){
    
    d <- roh_dat_incl_class %>% 
      filter(sample == s) %>% 
      # to transform to bed, substract 1bp to start only. end stays the same.
      mutate(start = start - 1) 
    
    breed <- unique(d$breed) %>% unique()
    
    fl_nm <- paste0(breed,"_",s,".bed")
    
    d %>% 
      dplyr::select(chr, start, end) %>% 
      arrange(chr, start) %>% 
      write.table(file = here("00_analysis/data/bed_files_ROH",fl_nm), sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
  })

```

## Import bedtools results
```{r import_bed_and_count_intersections}
roh_multint_dat <- here("12_LD_ROH/output/SOSFERT_ROH_intersections.bed") %>% 
  vroom(col_names = FALSE) %>% 
  dplyr::rename(seqnames=X1, start=X2, end=X3) %>% 
  # do not forget to convert back to 1-start-fully-closed format
  # just add 1bp to start
  mutate(start = start + 1) 
```

## For each intersection, count samples per breed and export

Computation is not quick, better do it once, then export.
```{r process_bed_and_count_intersections, eval = FALSE}

breed_cts <- roh_multint_dat$X5 %>% 
  str_split(",") %>% 
  lapply(function(x) str_sub(string = x, start = 1,end = 2)) %>% 
  lapply(function(x){
    data.frame(LW = sum(x == "LW"), PI = sum(x == "PI"))
  }) %>% 
  bind_rows()
  
roh_multint_dat2 <- bind_cols(
  
  roh_multint_dat %>% dplyr::select(seqnames, start, end),
  
  breed_cts
) 

write.csv(x = roh_multint_dat2, file = here("00_analysis/data/roh_intersections.csv"))

rm(roh_multint_dat, breed_cts, roh_multint_dat2)

```


```{r import_intersections}
roh_multint_dat2 <- here("00_analysis/data/roh_intersections.csv") %>% vroom() %>% dplyr::select(-"...1")
```

## Breed-private consensus ROH

```{r vis_bp_private_consensus_roh}

total_bp_private_LW <- roh_multint_dat2 %>% 
  filter(LW > 0, PI == 0) %>% 
  mutate(length = (end-start)+1) %>% 
  group_by(LW) %>% 
  summarise(bp_roh = sum(length)) %>% 
  dplyr::rename(n_samples = LW)
  
total_bp_private_PI <- roh_multint_dat2 %>% 
  filter(LW == 0, PI > 0) %>% 
  mutate(length = (end-start)+1) %>% 
  group_by(PI) %>% 
  summarise(bp_roh = sum(length)) %>% 
  dplyr::rename(n_samples = PI)

dat_vis_private <- list(bp_private_LW = total_bp_private_LW,
     bp_private_PI= total_bp_private_PI) %>% 
  bind_rows(.id="data_set") %>% 
  mutate(intersection_class = ifelse(n_samples <= 5, "1-5", "6-10")) %>% 
  reshape2::melt(id.vars = c("data_set", "n_samples", "bp_roh")) %>% 
  arrange(data_set, desc(n_samples)) %>% 
  group_by(data_set) %>% 
  mutate(cumsum = cumsum(bp_roh)) 

```

```{r tabulate_private_roh_bps_in_at_least_n_samples}

dat_vis_private %>% 
  reshape2::dcast(n_samples ~ data_set, value.var = "cumsum") %>% 
  dplyr::rename(min_no_samples = n_samples) %>% 
  kable(format.args = list(big.mark = ","), caption = "Total breed-private bp in ROH in at least n samples")

```

## Shared ROHs between LW and PI

```{r vis_shared_roh_bps_in_at_least_n_samples}

dat_vis_shared <- lapply(1:10, function(i){
  
  roh_multint_dat2 %>% 
    filter(LW == i & PI == i) %>% 
    mutate(n_samples = LW + PI,
           samples_per_breed = i) 
  
  }) %>% 
  bind_rows() %>% 
  group_by(samples_per_breed) %>% 
  mutate(length = (end-start)+1) %>% 
  summarise(total_bp = sum(length)) %>% 
  arrange(desc(samples_per_breed)) %>% 
  mutate(cumsum = cumsum(total_bp)) %>%
  mutate(intersection_class = ifelse(samples_per_breed <= 5, "1-5", "6-10"))
  
```

```{r tabulate_shared_roh_bps_in_at_least_n_samples}

dat_vis_shared %>% 
  arrange(samples_per_breed) %>% 
  dplyr::select(samples_per_breed, cumsum) %>% 
  dplyr::rename(min_no_samples_per_breed = samples_per_breed, total_bp = cumsum) %>% 
  kable(format.args = list(big.mark = ","), caption = "Total shared bp in ROH in at least n samples")

```

## Summary 

```{r vis_private_shared_roh, fig.height=9}

# private
p1 <- dat_vis_private %>% 
  filter(value == "1-5") %>% 
  ggplot(aes(x = n_samples, y = cumsum/1e6)) +
    geom_bar(stat = "identity") +
    facet_grid(data_set~value, scales = "free") +
    scale_x_continuous(breaks = scales::pretty_breaks(5)) +
    scale_y_continuous(breaks = scales::pretty_breaks(10)) +
    theme_bw(base_size = 10) +
    ylab("Mbp") +
    xlab("min. no. samples")


p2 <- dat_vis_private %>% 
  filter(value == "6-10") %>% 
  ggplot(aes(x = n_samples, y = cumsum/1e6)) +
    geom_bar(stat = "identity") +
    facet_grid(data_set~value, scales = "free") +
    scale_x_continuous(breaks = scales::pretty_breaks(5)) +
    scale_y_continuous(breaks = scales::pretty_breaks(10)) +
    theme_bw(base_size = 10) +
    ylab("Mbp") +
    xlab("min. no. samples")

# shared
p3 <- dat_vis_shared %>% 
  filter(intersection_class == "1-5") %>% 
  ggplot(aes(x = samples_per_breed, y = cumsum/1e6)) +
    geom_bar(stat = "identity") +
    facet_wrap(~intersection_class, scales = "free") +
    scale_x_continuous(breaks = scales::pretty_breaks(5)) +
    scale_y_continuous(breaks = scales::pretty_breaks(10)) +
    theme_bw(base_size = 10) +
    ylab("Mbp") +
    xlab("min. no. samples per breed")

p4 <- dat_vis_shared %>% 
  filter(intersection_class == "6-10") %>% 
  ggplot(aes(x = samples_per_breed, y = cumsum/1e6)) +
    geom_bar(stat = "identity") +
    facet_wrap(~intersection_class, scales = "free") +
    scale_x_continuous(breaks = scales::pretty_breaks(5)) +
    scale_y_continuous(breaks = scales::pretty_breaks(10)) +
    theme_bw(base_size = 10) +
    ylab("Mbp") +
    xlab("min. no. samples per breed")

p5 <- ggpubr::ggarrange(p1,p2,ncol = 2) %>% 
  ggpubr::annotate_figure(top = "Genome in private ROH")

p6 <- ggpubr::ggarrange(p3,p4,ncol = 2) %>% 
  ggpubr::annotate_figure(top = "Genome in shared ROH")

ggpubr::ggarrange(p5,p6, nrow = 2, heights = c(1.4,1))
```

# Check sample contribution at each level of consensus

Min. no. samples is expected to vary along the genome. Check the overall contribution of each sample at any minimum sample required.

## Contribution of samples to breed-specific ROHs
```{r bp_sample_contribution_private_LW}

cols <- c("seqnames", "start", "end", "LW_H29777", "LW_H29778", "LW_H29779", "LW_H29780", "LW_H29781", "LW_H29782", "LW_H29783", "LW_H29784", "LW_H29785", "LW_H29786", "PI_H29767", "PI_H29768", "PI_H29769", "PI_H29770", "PI_H29771", "PI_H29772", "PI_H29773", "PI_H29774", "PI_H29775", "PI_H29776")

roh_multint_dat %>% 
  dplyr::select(seqnames, start, end, X6:X25) %>% 
  magrittr::set_colnames(cols) %>% 
  inner_join(
    roh_multint_dat2 %>% filter(LW > 0, PI == 0),
    by = c("seqnames","start","end")
  ) %>% 
  reshape2::melt(id.vars = c("seqnames","start","end","LW","PI"), variable.name = "sample_id") %>% 
  as_tibble() %>% 
  filter(value > 0) %>% 
  group_by(LW, sample_id) %>% 
  mutate(length_bp = (end-start) + 1) %>% 
  summarise(total_Mbp = sum(length_bp)/1e6) %>% 
  mutate(sample_id = str_remove(sample_id, "LW_")) %>% 
  ggplot(aes(x = sample_id, y = total_Mbp)) +
    geom_bar(stat = "identity") +
    facet_wrap(~LW, scales = "free_y") +
    theme_bw(base_size = 12) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
    ggtitle("bp_sample_contribution_private_LW")

```

```{r bp_sample_contribution_private_PI}
roh_multint_dat %>% 
  dplyr::select(seqnames, start, end, X6:X25) %>% 
  magrittr::set_colnames(cols) %>% 
  inner_join(
    roh_multint_dat2 %>% filter(LW == 0, PI > 0),
    by = c("seqnames","start","end")
  ) %>% 
  reshape2::melt(id.vars = c("seqnames","start","end","LW","PI"), variable.name = "sample_id") %>% 
  as_tibble() %>% 
  filter(value > 0) %>% 
  group_by(PI, sample_id) %>% 
  mutate(length_bp = (end-start) + 1) %>% 
  summarise(total_Mbp = sum(length_bp)/1e6) %>% 
  mutate(sample_id = str_remove(sample_id, "PI_")) %>% 
  ggplot(aes(x = sample_id, y = total_Mbp)) +
    geom_bar(stat = "identity") +
    facet_wrap(~PI, scales = "free_y") +
    theme_bw(base_size = 12) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
    ggtitle("bp_sample_contribution_private_PI")
```

```{r bp_sample_contribution_shared, fig.width=15}

roh_multint_dat %>% 
  dplyr::select(seqnames, start, end, X6:X25) %>% 
  magrittr::set_colnames(cols) %>% 
  inner_join(
    
    lapply(1:10, function(i){
      roh_multint_dat2 %>% 
        filter(LW == i & PI == i) %>% 
        mutate(n_samples = LW + PI,
               samples_per_breed = i) 
      }) %>% 
      bind_rows() %>% 
      dplyr::select(-n_samples, -LW, -PI),
    
    by = c("seqnames","start","end")
    
  ) %>% 
  reshape2::melt(id.vars = c("seqnames","start","end","samples_per_breed"), variable.name = "sample_id") %>% 
  as_tibble() %>% 
  filter(value > 0) %>% 
  group_by(samples_per_breed, sample_id) %>% 
  mutate(length_bp = (end-start) + 1) %>% 
  summarise(total_Mbp = sum(length_bp)/1e6) %>% 
  mutate(sample_id = str_remove(sample_id, "PI_"),
         sample_id = str_remove(sample_id, "LW_")) %>% 
  ggplot(aes(x = sample_id, y = total_Mbp)) +
    geom_bar(stat = "identity") +
    facet_wrap(~samples_per_breed, scales = "free_y") +
    theme_bw(base_size = 12) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
    ggtitle("bp_sample_contribution_shared")

```



# Consensus from here on now: `r minN <- 8` `r minN`/10 samples.

Consensus: If `r minN` out 10 samples are consistently sharing ROHs.


```{r set_consensus_data_sets}

private_LW_minN <- roh_multint_dat2 %>% filter(LW >= minN, PI == 0) 
  
private_PI_minN <- roh_multint_dat2 %>% filter(LW == 0, PI >= minN) 

shared_LWminN_PIminN <- roh_multint_dat2 %>% filter(LW >= minN & PI >= minN)

```

Some consensus ROHs can occur next to each other. Merge those.

```{r merge_rohs}

private_LW_minN_merged <- private_LW_minN %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>% 
  reduce(min.gapwidth = 1)

private_PI_minN_merged <- private_PI_minN %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>% 
  reduce(min.gapwidth = 1)

shared_LWminN_PIminN_merged <- shared_LWminN_PIminN %>% 
  makeGRangesFromDataFrame(keep.extra.columns = TRUE) %>% 
  reduce(min.gapwidth = 1)

```

## How many ROH-bases are private vs shared?

```{r bp_total_private_shared}

list(
  private_LW = private_LW_minN_merged %>% as_tibble(),
  private_PI = private_PI_minN_merged %>% as_tibble(),
  shared = shared_LWminN_PIminN_merged %>% as_tibble()
  ) %>% 
  bind_rows(.id = "data_set") %>% 
  group_by(data_set) %>% 
  mutate(len_bp = (end-start)+1) %>% 
  summarise(sum(len_bp)) %>% 
  kable(format.args = list(big.mark = ","))

```

There are much more shared roh-bases than private roh-bases.


## What proportion of ROHs are below 100bp? - less than 100bp might too low.

```{r roh_size_distribution}

list(
  private_LW = private_LW_minN_merged %>% as_tibble(),
  private_PI = private_PI_minN_merged %>% as_tibble(),
  shared = shared_LWminN_PIminN_merged %>% as_tibble()
  ) %>% 
  bind_rows(.id = "data_set") %>% 
  group_by(data_set) %>% 
  mutate(len_bp = (end-start)+1, is_too_short = len_bp < 100) %>% 
  summarise(total_too_short = sum(is_too_short), pct_too_short = mean(is_too_short)*100, min_len = min(len_bp)) %>% 
  kable(digits = 2)

```

There are not too many "too short" ROHs, so I will just remove them.

```{r exclude_too_short_rohs}

private_LW_minN_merged_GTE100bp <- private_LW_minN_merged %>% as_tibble() %>% mutate(len_bp = (end-start)+1) %>% filter(len_bp >= 100) %>% makeGRangesFromDataFrame(keep.extra.columns = TRUE)

private_PI_minN_merged_GTE100bp <- private_PI_minN_merged %>% as_tibble() %>% mutate(len_bp = (end-start)+1) %>% filter(len_bp >= 100) %>% makeGRangesFromDataFrame(keep.extra.columns = TRUE)

shared_LWminN_PIminN_merged_GTE100bp <- shared_LWminN_PIminN_merged %>% as_tibble() %>% mutate(len_bp = (end-start)+1) %>% filter(len_bp >= 100) %>% makeGRangesFromDataFrame(keep.extra.columns = TRUE)

```

# Get genes for private and shared ROHs

```{r gene_set_scc}

ssc_gene_set <- readRDS(here("ref_ensemble95/gene_set/ssc_gene_set.rds"))

```

```{r genes_in_rohs}

private_LW_minN_merged_GTE100bp_genes <- subsetByOverlaps(ssc_gene_set, private_LW_minN_merged_GTE100bp, minoverlap = 1)

private_PI_minN_merged_GTE100bp_genes <- subsetByOverlaps(ssc_gene_set, private_PI_minN_merged_GTE100bp, minoverlap = 1)

shared_LWminN_PIminN_merged_GTE100bp_genes <- subsetByOverlaps(ssc_gene_set, shared_LWminN_PIminN_merged_GTE100bp, minoverlap = 1)

```

# Summarize gene ontologies (GO-BP)

The purpose of this section is to obtain an idea of the biological functions of genes in private and shared ROHs.

## Obtain gene ontologies

Do this one time and export.

```{r setup_biomart, eval = FALSE}

#ensembl <- useEnsembl(biomart = "genes", dataset = "sscrofa_gene_ensembl", version = "95")
ensembl <- useEnsembl(biomart = "genes", dataset = "sscrofa_gene_ensembl", version = "95")

get_go <- function(dd){
  
  #dd <- private_LW_minN_merged_GTE100bp_genes
  
  # get gene ontology ids
  q <- getBM(attributes = c("ensembl_gene_id", "go_id", "name_1006", "namespace_1003", "external_gene_name"), 
             filters = "ensembl_gene_id", 
             values = dd$gene_id, 
             mart = ensembl)
  
  res <- q %>% 
    filter(namespace_1003 == "biological_process") %>% # retain GO_BPs
    mutate(external_gene_name = ifelse(external_gene_name == "", NA, external_gene_name),
           ensid_sym = paste0(ensembl_gene_id, " (",external_gene_name, ")")) %>% 
    as_tibble() %>% 
    dplyr::select(name_1006, ensid_sym, go_id) %>% 
    group_by(go_id, name_1006) %>% 
    summarise(n_genes=n(), genes = paste0(ensid_sym, collapse = "; ")) %>% # count no. genes per GO-BP
    arrange(desc(n_genes)) %>% 
    dplyr::rename(go_name = name_1006)
  
  return(res)

}

```

```{r get_ontologies_and_export, eval = FALSE}

private_LW_minN_merged_GTE100bp_genes_GOBP <- get_go(dd = private_LW_minN_merged_GTE100bp_genes)
private_PI_minN_merged_GTE100bp_genes_GOBP <- get_go(dd = private_PI_minN_merged_GTE100bp_genes)
shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP <- get_go(dd = shared_LWminN_PIminN_merged_GTE100bp_genes)

write.csv(private_LW_minN_merged_GTE100bp_genes_GOBP,
          here("00_analysis/data/ROH_GOBP","private_LW_minN_merged_GTE100bp_genes_GOBP.csv"))

write.csv(private_PI_minN_merged_GTE100bp_genes_GOBP,
          here("00_analysis/data/ROH_GOBP","private_PI_minN_merged_GTE100bp_genes_GOBP.csv"))

write.csv(shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP,
          here("00_analysis/data/ROH_GOBP","shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP.csv"))

rm(private_LW_minN_merged_GTE100bp_genes_GOBP, private_PI_minN_merged_GTE100bp_genes_GOBP, shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP)
```

```{r import_GOBP_annotations}

private_LW_minN_merged_GTE100bp_genes_GOBP <- here("00_analysis/data/ROH_GOBP","private_LW_minN_merged_GTE100bp_genes_GOBP.csv") %>% vroom()

private_PI_minN_merged_GTE100bp_genes_GOBP <- here("00_analysis/data/ROH_GOBP","private_PI_minN_merged_GTE100bp_genes_GOBP.csv") %>% vroom()

shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP <- here("00_analysis/data/ROH_GOBP","shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP.csv") %>% vroom()
```


## Summarize ontologies with rrvgo

Get semantic similarities, reduce terms and export
```{r prep_rrvgo_dat, eval = FALSE}

prep_rrvgo <- function(d){
  
  #d <- private_LW_minN_merged_GTE100bp_genes_GOBP
  
  sim_ma <- calculateSimMatrix(x = d$go_id, orgdb="org.Ss.eg.db", ont="BP", method="Rel")
  
  scores <- setNames(d$n_genes, d$go_id)
  
  reduced_terms <- reduceSimMatrix(sim_ma, scores, threshold=0.7, orgdb="org.Ss.eg.db")
  
  res <- list(sim_ma, reduced_terms)
  
  names(res) <- c("sim_ma","reduced_terms")
  
  return(res)
  
  }

private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo <- prep_rrvgo(private_LW_minN_merged_GTE100bp_genes_GOBP)
private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo <- prep_rrvgo(private_PI_minN_merged_GTE100bp_genes_GOBP)
shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo <- prep_rrvgo(shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP)

saveRDS(private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo, 
        here("00_analysis/data/rrvgo/private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo.rds"))
saveRDS(private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo, 
        here("00_analysis/data/rrvgo/private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo.rds"))
saveRDS(shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo, 
        here("00_analysis/data/rrvgo/shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo.rds"))

rm(private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo, private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo, shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo)
```

```{r import_rrvgo_res}

private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo <- readRDS(
  here("00_analysis/data/rrvgo/private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo.rds")
  )

private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo <- readRDS(
  here("00_analysis/data/rrvgo/private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo.rds")
  )

shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo <- readRDS(
  here("00_analysis/data/rrvgo/shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo.rds")
  )
```


### Large White private ROH - Gene Annotation (GO-BP)

#### Tree map plot
```{r fig.width=12, fig.height=8}

treemapPlot(private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms)

```

#### PCA semmantic simmilarity
```{r fig.width=12, fig.height=12}

scatterPlot(simMatrix = private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo$sim_ma, 
            reducedTerms = private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms)

```

#### Reduced Terms
```{r}
private_LW_minN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms %>% 
  group_by(parentTerm) %>% 
  summarise(n_terms = n(), terms = paste(term, collapse = "; ")) %>% 
  arrange(desc(n_terms)) %>% 
  DT::datatable(rownames = FALSE)
```

#### Genes by term
```{r}
private_LW_minN_merged_GTE100bp_genes_GOBP %>% 
  dplyr::select(-"...1") %>% 
  DT::datatable(rownames = FALSE)
```

### Pietrain private ROH - Gene Annotation (GO-BP)

#### Tree map plot
```{r fig.width=12, fig.height=8}

treemapPlot(private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms)

```

#### PCA semmantic simmilarity
```{r fig.width=12, fig.height=12}

scatterPlot(simMatrix = private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo$sim_ma, 
            reducedTerms = private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms)

```

#### Reduced Terms
```{r}
private_PI_minN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms %>% 
  group_by(parentTerm) %>% 
  summarise(n_terms = n(), terms = paste(term, collapse = "; ")) %>% 
  arrange(desc(n_terms)) %>% 
  DT::datatable(rownames = FALSE)
```

#### Genes by term
```{r}
private_PI_minN_merged_GTE100bp_genes_GOBP %>% 
  dplyr::select(-"...1") %>% 
  DT::datatable(rownames = FALSE)
```

### Shared ROH - Gene Annotation (GO-BP)

#### Tree map plot
```{r fig.width=12, fig.height=8}

treemapPlot(shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms)

```

#### PCA semmantic simmilarity
```{r fig.width=12, fig.height=12}

scatterPlot(simMatrix = shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo$sim_ma, 
            reducedTerms = shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms)

```

#### Reduced Terms
```{r}
shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP_rrvgo$reduced_terms %>% 
  group_by(parentTerm) %>% 
  summarise(n_terms = n(), terms = paste(term, collapse = "; ")) %>% 
  arrange(desc(n_terms)) %>% 
  DT::datatable(rownames = FALSE)
```

#### Genes by term
```{r}
shared_LWminN_PIminN_merged_GTE100bp_genes_GOBP %>% 
  dplyr::select(-"...1") %>% 
  DT::datatable(rownames = FALSE)
```


# What type of annotations do SNPs contained in ROHs have?

## Import SNP set
```{r eval = FALSE}

# this takes a while, so run it once and then export it
snps_in_ROH <- subsetByOverlaps(
  
  x = here("10_FinalVCF/output", "cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered.frq") %>% 
    vroom() %>% 
    dplyr::select(CHROM, POS) %>% 
    mutate(seqnames = CHROM, start = POS, end = POS) %>% 
    dplyr::select(seqnames, start, end) %>% 
    makeGRangesFromDataFrame(), 
  
 ranges = c(private_LW_minN_merged_GTE100bp, 
            private_PI_minN_merged_GTE100bp, 
            shared_LWminN_PIminN_merged_GTE100bp), 
 
 minoverlap = 1
 )

saveRDS(snps_in_ROH, here("00_analysis/data/26_ROH_SOSFERT_samples", "snps_in_ROH.rds"))

rm(snps_in_ROH)
```

```{r}

snps_in_ROH <- here("00_analysis/data/26_ROH_SOSFERT_samples", "snps_in_ROH.rds") %>% readRDS()

```

Out of the 16M, ``r length(snps_in_ROH)`` SNPs are found in (private, shared) consensus ROHs.

## How many of these SNPs are actualy fixed?
```{r eval = FALSE}

# it takes a while, expot it
snps_in_ROH_with_freq <- inner_join(
  snps_in_ROH %>% as_tibble(),
  here("10_FinalVCF/output", "cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered.frq") %>% 
    vroom(), 
  by = c("seqnames"="CHROM", "start" = "POS")
  ) %>% 
  dplyr::select(seqnames, start, end, `{ALLELE:FREQ}`) %>% 
  separate(col = `{ALLELE:FREQ}`, into = c("ref_freq_tmp","alt_freq_tmp"), sep = c("\t")) %>% 
  separate(col = ref_freq_tmp, into = c("ref","ref_freq"), sep = c(":")) %>% 
  separate(col = alt_freq_tmp, into = c("alt","alt_freq"), sep = c(":"))
           
saveRDS(snps_in_ROH_with_freq, here("00_analysis/data/26_ROH_SOSFERT_samples", "snps_in_ROH_with_freq.rds"))

rm(snps_in_ROH_with_freq)
```

```{r}
snps_in_ROH_with_freq <- here("00_analysis/data/26_ROH_SOSFERT_samples", "snps_in_ROH_with_freq.rds") %>% readRDS()

snps_in_ROH_with_freq_fixed <- snps_in_ROH_with_freq %>% filter(ref_freq == 0 | ref_freq == 1)
```

Out of the ``r length(snps_in_ROH)``, ``r nrow(snps_in_ROH_with_freq_fixed)`` are fixed.

## Check annotations for fixed SNPs in ROHs of interest
```{r}

# takes too long, export it
snps_in_ROH_with_freq_fixed_annot <- 
  
  inner_join(
    
    here("16_sNPeff/output","cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered.ann.tab") %>% 
      vroom(),
    
    snps_in_ROH_with_freq_fixed,
    
    by = c("CHROM"="seqnames", "POS"="start")
    
    )

saveRDS(snps_in_ROH_with_freq_fixed_annot, here("00_analysis/data/26_ROH_SOSFERT_samples", "snps_in_ROH_with_freq_fixed_annot.rds"))

rm(snps_in_ROH_with_freq_fixed_annot)
```

```{r}
snps_in_ROH_with_freq_fixed_annot <- here("00_analysis/data/26_ROH_SOSFERT_samples", "snps_in_ROH_with_freq_fixed_annot.rds") %>% 
  readRDS()
```

```{r}
snps_in_ROH_with_freq_fixed_annot %>% 
  group_by(`ANN[*].IMPACT`) %>% 
  summarise(n=n())
```

As it is usually the case, only very few SNPs have a high impact and most SNPs are belong to the "MODIFIER" category.

Inspect SNPs classified as having high or moderate impacts:
```{r}
snps_in_ROH_with_freq_fixed_annot %>% 
  filter(`ANN[*].IMPACT` %in% c("HIGH","MODERATE")) %>% 
  DT::datatable()
```

Get Gene ontologies (BP) for those genes (compare with sections below):
```{r eval = FALSE}
# run once and export
ensembl <- useEnsembl(biomart = "genes", dataset = "sscrofa_gene_ensembl", version = "95")

tmp_query <- getBM(attributes = c("ensembl_gene_id", "go_id", "name_1006", "namespace_1003", "external_gene_name"), 
             filters = "ensembl_gene_id", 
             values = snps_in_ROH_with_freq_fixed_annot %>% 
                        filter(`ANN[*].IMPACT` %in% c("HIGH","MODERATE")) %>% 
                        .$`ANN[*].GENEID` %>% 
                        unique(), 
             mart = ensembl) 

saveRDS(tmp_query, here("00_analysis/data/26_ROH_SOSFERT_samples","moderate_high_impact_ROH_fixed_SNPs.rds"))

```

```{r}

here("00_analysis/data/26_ROH_SOSFERT_samples","moderate_high_impact_ROH_fixed_SNPs.rds") %>%
  readRDS() %>% 
  filter(namespace_1003 == "biological_process") %>% 
  DT::datatable()

```


## How are moderate/high impact SNPs distributed among ROHs?

As it can be seen below, private-LW ROHs is the only set that do not contain any impact SNPs.

High impact SNPs only occur in private Pietrain.They do not have GO-BPs.

GO-BPs only are found in the shared subset for two genes HAO2 and STON2. Interesntingly HAO2 is inolved in lipid metabolism.

### Private LW
```{r}

subsetByOverlaps(
  snps_in_ROH_with_freq_fixed_annot %>% 
    filter(`ANN[*].IMPACT` %in% c("HIGH","MODERATE")) %>% 
    mutate(seqnames = CHROM, start = POS, end = POS) %>% 
    dplyr::select(-CHROM, -POS) %>% 
    makeGRangesFromDataFrame(keep.extra.columns = TRUE),
  private_LW_minN_merged_GTE100bp,
  minoverlap = 1
) %>% 
  as_tibble()


```

### Private PI
```{r}
subsetByOverlaps(
  snps_in_ROH_with_freq_fixed_annot %>% 
    filter(`ANN[*].IMPACT` %in% c("HIGH","MODERATE")) %>% 
    mutate(seqnames = CHROM, start = POS, end = POS) %>% 
    dplyr::select(-CHROM, -POS) %>% 
    makeGRangesFromDataFrame(keep.extra.columns = TRUE),
  private_PI_minN_merged_GTE100bp,
  minoverlap = 1
) %>% 
  as_tibble() %>% 
  DT::datatable()
```

### Shared
```{r}
subsetByOverlaps(
  snps_in_ROH_with_freq_fixed_annot %>% 
    filter(`ANN[*].IMPACT` %in% c("HIGH","MODERATE")) %>% 
    mutate(seqnames = CHROM, start = POS, end = POS) %>% 
    dplyr::select(-CHROM, -POS) %>% 
    makeGRangesFromDataFrame(keep.extra.columns = TRUE),
  shared_LWminN_PIminN_merged_GTE100bp,
  minoverlap = 1
) %>% 
  as_tibble() %>% 
  DT::datatable()
```

