---
title: "Genome analysis dam vs sire lines"
output: 
  html_document:
    toc: 5
---

```{r libraries, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, include = TRUE, message = FALSE)

library(vroom)
library(tidyverse)
library(GenomicRanges)
library(here)
library(knitr)

```

```{r sample_info, include=FALSE}
sample_info <- vroom(here("sample_info/internal_external_sample_info.tsv"))

```

```{r}
genome_length <- 2501912388 #Golden Path Length	2,501,912,388 (Sscrofa11.1)

```



# Intro

Runs of homozygosity were computed for each sample of the SOSFERT sample set, using bcftools.

These samples were sequenced at high coverage of around 50x, so the VCF file had ~15M SNPs. These were used to compute ROH.

For more details check the corresponding script: in ~/12_LD_ROH/scripts/12_bcftools_ROH_SOSFERT_vcf.sh

```{r ROH_import_regions, include=FALSE}

roh_dat <- vroom(here("12_LD_ROH/output/cohort_biallelicSNPs_HardFiltered_WithMissingness_SOSFERTsamples_GTfiltered.roh.table"), 
                 col_types = c(`[3]Chromosome`="c")) %>% 
  dplyr::rename(RG = `# RG`, sample = `[2]Sample`, chr = `[3]Chromosome`, start = `[4]Start`, end = `[5]End`, length_bp = `[6]Length (bp)`,
                n_markers = `[7]Number of markers`, qual_avg_fwd_bwd_phred = `[8]Quality (average fwd-bwd phred score)`) %>% 
  filter(RG != "# RG") %>% 
  # exclude x-chrom
  filter(chr != "X") %>% 
  inner_join(
    
    # include an ordered-naming for later when plotting all samples together (i.e. boxplot)
    sample_info %>% 
      filter(project == "SOSFERT") %>% 
      group_by(breed) %>% 
      mutate(breed = ifelse(breed == "Pietrain", "PI", breed),
             breed = ifelse(breed == "Large White", "LW", breed)) %>% 
      arrange(breed, sample_id), 

    by = c("sample" = "sample_id")
    )


dim(roh_dat) # 100140     10
```


# ROH length distribution by ROH-class
Based on: "ROHs were separated into three size classes: 1) small (<100 Kbp), 2) medium (0.1 to 5 Mbp) and 3) large (>5 Mbp). "
PLoS Genet. 2012;8(11):e1003100. doi: 10.1371/journal.pgen.1003100. Epub 2012 Nov 29. Regions of homozygosity in the porcine genome: consequence of demography and the recombination landscape

```{r classify_rohs}

roh_dat_incl_class <- roh_dat %>% 
  mutate(roh_class = ifelse(length_bp < 1e5, "short", NA),
         roh_class = ifelse(length_bp >= 1e5 & length_bp <= 5e6, "medium", roh_class),
         roh_class = ifelse(length_bp > 5e6, "large", roh_class),
         roh_class = factor(roh_class, levels = c("short", "medium", "large")))

```


```{r roh_length_boxplots, fig.width=10}

roh_dat_incl_class %>% 
  ggplot(aes(x = length_bp, y = sample)) +
    geom_boxplot() +
    facet_grid(breed ~ roh_class, scales = "free") +
    theme_bw() +
    scale_x_continuous(breaks = scales::pretty_breaks(5)) +
    theme(axis.text.x = element_text(angle = 45, hjust = 0.5, vjust = 0.5))

```

# Number of roh per class in each sample

```{r summarise_roh_counts}

roh_dat_incl_class %>% 
  group_by(sample,breed,roh_class) %>% 
  summarise(n=n()) %>% 
  reshape2::dcast(breed+sample ~ roh_class, value.var = "n") %>% 
  kable()
  
```

# Agreggate lengths (bp) of roh per class in each sample
```{r summarise_roh_aggreg_len}

roh_dat_incl_class %>% 
  group_by(sample,breed,roh_class) %>% 
  summarise(total_length=sum(length_bp)) %>% 
  reshape2::dcast(breed+sample ~ roh_class, value.var = "total_length") %>% 
  kable(format.args = list(big.mark = ","))

```

# Total fraction of genome in ROH
```{r}

roh_dat_incl_class %>% 
  group_by(sample,breed) %>% 
  summarise(total_length=sum(length_bp)) %>% 
  mutate(total_genome_fraction = total_length/genome_length) %>% 
  kable(format.args = list(big.mark = ","), digits = 2)

```


# Genome fractions in ROH by class
```{r per_sample_avg_genome_prop_as_roh} 

# get genomic fraction
roh_dat_incl_class_genome_fraction <- roh_dat_incl_class %>% 
  mutate(genome_fraction = length_bp/genome_length)

# get total fraction per sample
total <- roh_dat_incl_class_genome_fraction %>% 
  group_by(breed, sample) %>% 
  summarise(total = sum(genome_fraction)) %>% 
  mutate(roh_class = "total") %>% 
  dplyr::select(breed, sample, roh_class, total)

# get total fraction by class per sample
total_by_class <- roh_dat_incl_class_genome_fraction %>% 
  group_by(breed, sample, roh_class) %>% 
  summarise(total = sum(genome_fraction))

# combine data
fractions <- bind_rows(total, total_by_class) %>% 
  #summarise by breed and class
  group_by(breed, roh_class) %>% 
  summarise(n(), mean = mean(total), sd = sd(total)) %>% 
  # arrange levels for plot
  mutate(roh_class = factor(roh_class, levels = c("total","large","medium","short")))


errors_lower <- fractions$mean - fractions$sd
errors_upper <- fractions$mean + fractions$sd

ggplot(fractions, aes(x = breed, y = mean, fill = roh_class)) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_errorbar(aes(ymin = errors_lower, ymax = errors_upper), position =  position_dodge2(width = 0.5, padding = 0.5)) +
  theme_bw(base_size = 12) +
  xlab(NULL) +
  ylab("genome fraction") +
  ggtitle("Mean per-sample genome fractions as ROHs") +
  scale_fill_brewer(palette="Dark2") +
  scale_y_continuous(breaks = seq(0,1,0.05), expand = c(0.01,0.01)) +
  theme(legend.title = element_blank(), plot.title = element_text(hjust = 0.5))

```




```{r}

roh_uniq <- roh_dat_incl_class %>% dplyr::select(chr, start, end) %>% unique() %>% makeGRangesFromDataFrame()


gr <- roh_dat_incl_class %>% filter(sample == "H29777") %>% dplyr::select(chr, start, end) %>% makeGRangesFromDataFrame()


ov <- findOverlaps(query = gr, subject = roh_uniq, minoverlap = 1)

```


https://support.bioconductor.org/p/72656/

https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-021-07992-6#Sec14
